File: internal/schema/schema.go
// Package schema defines the structural QUALITY.md frontmatter schema.
package schema

// NodeKind identifies a structural frontmatter node.
type NodeKind string

const (
	ModelKind       NodeKind = "model"
	TargetKind      NodeKind = "target"
	FactorKind      NodeKind = "factor"
	RequirementKind NodeKind = "requirement"
	RatingLevelKind NodeKind = "ratingLevel"
)

const (
	PropertyTitle        = "title"
	PropertyRatingScale  = "ratingScale"
	PropertyFactors      = "factors"
	PropertyRequirements = "requirements"
	PropertyTargets      = "targets"
	PropertySource       = "source"
	PropertyLevel        = "level"
	PropertyDescription  = "description"
	PropertyCriterion    = "criterion"
	PropertyAssessment   = "assessment"
	PropertyRatings      = "ratings"
)

// Shape is the YAML shape a property value must have.
type Shape string

const (
	ScalarShape   Shape = "scalar"
	MapShape      Shape = "map"
	SequenceShape Shape = "sequence"
)

// Presence describes whether a property is required, recommended, or optional.
type Presence string

const (
	RequiredPresence    Presence = "required"
	RecommendedPresence Presence = "recommended"
	OptionalPresence    Presence = "optional"
)

// Property is one valid key on a structural node.
type Property struct {
	Name         string
	Shape        Shape
	Presence     Presence
	ElementKind  NodeKind
	ElementShape Shape
	ValueShape   Shape
	MinItems     int
}

// RequiredAny requires at least one of its properties to contain entries.
type RequiredAny struct {
	Name       string
	Properties []string
}

// Node is the structural schema for a frontmatter node.
type Node struct {
	Kind        NodeKind
	Properties  []Property
	RequiredAny []RequiredAny
}

// Property returns the schema property named name.
func (n Node) Property(name string) (Property, bool) {
	for _, property := range n.Properties {
		if property.Name == name {
			return property, true
		}
	}
	return Property{}, false
}

// HasProperty reports whether name is valid on the node.
func (n Node) HasProperty(name string) bool {
	_, ok := n.Property(name)
	return ok
}

// PropertyNames returns the valid property names in schema order.
func (n Node) PropertyNames() []string {
	names := make([]string, 0, len(n.Properties))
	for _, property := range n.Properties {
		names = append(names, property.Name)
	}
	return names
}

// Model is the root QUALITY.md model schema.
var Model = Node{
	Kind: ModelKind,
	Properties: []Property{
		{Name: PropertyTitle, Shape: ScalarShape, Presence: RecommendedPresence},
		{Name: PropertyRatingScale, Shape: SequenceShape, Presence: RequiredPresence, ElementKind: RatingLevelKind, MinItems: 2},
		{Name: PropertyFactors, Shape: MapShape, Presence: OptionalPresence, ElementKind: FactorKind},
		{Name: PropertyRequirements, Shape: MapShape, Presence: OptionalPresence, ElementKind: RequirementKind},
		{Name: PropertyTargets, Shape: MapShape, Presence: OptionalPresence, ElementKind: TargetKind},
		{Name: PropertySource, Shape: ScalarShape, Presence: OptionalPresence},
	},
	RequiredAny: []RequiredAny{
		{Name: "model-content", Properties: []string{PropertyFactors, PropertyRequirements, PropertyTargets}},
	},
}

// Target is the recursive target-node schema.
var Target = Node{
	Kind: TargetKind,
	Properties: []Property{
		{Name: PropertyFactors, Shape: MapShape, Presence: OptionalPresence, ElementKind: FactorKind},
		{Name: PropertyRequirements, Shape: MapShape, Presence: OptionalPresence, ElementKind: RequirementKind},
		{Name: PropertyTargets, Shape: MapShape, Presence: OptionalPresence, ElementKind: TargetKind},
		{Name: PropertySource, Shape: ScalarShape, Presence: OptionalPresence},
	},
}

// Factor is the recursive factor-node schema.
var Factor = Node{
	Kind: FactorKind,
	Properties: []Property{
		{Name: PropertyDescription, Shape: ScalarShape, Presence: RecommendedPresence},
		{Name: PropertyFactors, Shape: MapShape, Presence: OptionalPresence, ElementKind: FactorKind},
		{Name: PropertyRequirements, Shape: MapShape, Presence: OptionalPresence, ElementKind: RequirementKind},
	},
}

// Requirement is the requirement-node schema.
var Requirement = Node{
	Kind: RequirementKind,
	Properties: []Property{
		{Name: PropertyAssessment, Shape: ScalarShape, Presence: RequiredPresence},
		{Name: PropertyFactors, Shape: SequenceShape, Presence: OptionalPresence, ElementShape: ScalarShape},
		{Name: PropertyRatings, Shape: MapShape, Presence: OptionalPresence, ValueShape: ScalarShape},
	},
}

// RatingLevel is one rating-scale level schema.
var RatingLevel = Node{
	Kind: RatingLevelKind,
	Properties: []Property{
		{Name: PropertyLevel, Shape: ScalarShape, Presence: RequiredPresence},
		{Name: PropertyTitle, Shape: ScalarShape, Presence: OptionalPresence},
		{Name: PropertyDescription, Shape: ScalarShape, Presence: RecommendedPresence},
		{Name: PropertyCriterion, Shape: ScalarShape, Presence: RequiredPresence},
	},
}

// Nodes lists every structural schema node.
var Nodes = []Node{Model, Target, Factor, Requirement, RatingLevel}

File: src/domain/model/schema.ts
export const ModelNamePattern = /^[A-Za-z0-9](?:[A-Za-z0-9_-]*[A-Za-z0-9])?$/
export const ModelNamePatternText = "^[A-Za-z0-9](?:[A-Za-z0-9_-]*[A-Za-z0-9])?$"

export type NodeKind = "model" | "area" | "factor" | "requirement" | "ratingLevel"
export type Shape = "scalar" | "map" | "sequence"
export type Presence = "required" | "recommended" | "optional"

export interface Property {
  readonly name: string
  readonly shape: Shape
  readonly presence: Presence
  readonly elementKind?: NodeKind
  readonly elementShape?: Shape
  readonly valueShape?: Shape
  readonly minItems?: number
  readonly keyPattern?: string
  readonly pattern?: string
}

export interface SchemaNode {
  readonly kind: NodeKind
  readonly properties: ReadonlyArray<Property>
  readonly requiredAny?: ReadonlyArray<{
    readonly name: string
    readonly properties: ReadonlyArray<string>
  }>
}

export const property = {
  title: "title",
  ratingScale: "ratingScale",
  factors: "factors",
  requirements: "requirements",
  areas: "areas",
  source: "source",
  level: "level",
  description: "description",
  criterion: "criterion",
  assessment: "assessment",
  ratings: "ratings",
} as const

export const Model: SchemaNode = {
  kind: "model",
  properties: [
    { name: property.title, shape: "scalar", presence: "required" },
    { name: property.description, shape: "scalar", presence: "optional" },
    {
      name: property.ratingScale,
      shape: "sequence",
      presence: "required",
      elementKind: "ratingLevel",
      minItems: 2,
    },
    {
      name: property.factors,
      shape: "map",
      presence: "optional",
      elementKind: "factor",
      keyPattern: ModelNamePatternText,
    },
    {
      name: property.requirements,
      shape: "map",
      presence: "optional",
      elementKind: "requirement",
      keyPattern: ModelNamePatternText,
    },
    {
      name: property.areas,
      shape: "map",
      presence: "optional",
      elementKind: "area",
      keyPattern: ModelNamePatternText,
    },
    { name: property.source, shape: "scalar", presence: "optional" },
  ],
  requiredAny: [
    {
      name: "model-content",
      properties: [property.factors, property.requirements, property.areas],
    },
  ],
}

export const Area: SchemaNode = {
  kind: "area",
  properties: [
    { name: property.title, shape: "scalar", presence: "required" },
    { name: property.description, shape: "scalar", presence: "optional" },
    {
      name: property.factors,
      shape: "map",
      presence: "optional",
      elementKind: "factor",
      keyPattern: ModelNamePatternText,
    },
    {
      name: property.requirements,
      shape: "map",
      presence: "optional",
      elementKind: "requirement",
      keyPattern: ModelNamePatternText,
    },
    {
      name: property.areas,
      shape: "map",
      presence: "optional",
      elementKind: "area",
      keyPattern: ModelNamePatternText,
    },
    { name: property.source, shape: "scalar", presence: "optional" },
  ],
}

export const Factor: SchemaNode = {
  kind: "factor",
  properties: [
    { name: property.title, shape: "scalar", presence: "required" },
    { name: property.description, shape: "scalar", presence: "recommended" },
    {
      name: property.factors,
      shape: "map",
      presence: "optional",
      elementKind: "factor",
      keyPattern: ModelNamePatternText,
    },
    {
      name: property.requirements,
      shape: "map",
      presence: "optional",
      elementKind: "requirement",
      keyPattern: ModelNamePatternText,
    },
  ],
}

export const Requirement: SchemaNode = {
  kind: "requirement",
  properties: [
    { name: property.title, shape: "scalar", presence: "required" },
    { name: property.description, shape: "scalar", presence: "optional" },
    { name: property.assessment, shape: "scalar", presence: "required" },
    {
      name: property.factors,
      shape: "sequence",
      presence: "optional",
      elementShape: "scalar",
    },
    { name: property.ratings, shape: "map", presence: "optional", valueShape: "scalar" },
  ],
}

export const RatingLevel: SchemaNode = {
  kind: "ratingLevel",
  properties: [
    {
      name: property.level,
      shape: "scalar",
      presence: "required",
      pattern: ModelNamePatternText,
    },
    { name: property.title, shape: "scalar", presence: "required" },
    { name: property.description, shape: "scalar", presence: "recommended" },
    { name: property.criterion, shape: "scalar", presence: "required" },
  ],
}

export const schemaProperty = (node: SchemaNode, name: string) =>
  node.properties.find((candidate) => candidate.name === name)

export const hasSchemaProperty = (node: SchemaNode, name: string) =>
  schemaProperty(node, name) !== undefined
